source/fuzz/transformation_changing_memory_scope.cpp

```diff
@@ -0,0 +1,193 @@
+                                         ->GetSingleWordInOperand(0);
+
+  if (value_instruction_int_width != 32) {
+    return false;
+  }
+
+  // The new value of memory scope must available be wider than the older.
+  auto memory_scope_in_operand =
+      GetMemoryScopeInOperandIndex(needed_instruction->opcode());
+  auto new_memory_scope_value =
+      static_cast<SpvScope>(value_instruction->GetSingleWordInOperand(0));
+  auto old_memory_scope_value = static_cast<SpvScope>(
+      ir_context->get_def_use_mgr()
+          ->GetDef(needed_instruction->GetSingleWordInOperand(
+              memory_scope_in_operand))
+          ->GetSingleWordInOperand(0));
+
+  if (!IsValidScope(new_memory_scope_value, old_memory_scope_value)) {
+    return false;
+  }
+
+  return true;
+}
+
+void TransformationChangingMemoryScope::Apply(
+    opt::IRContext* ir_context, TransformationContext* /*unused*/) const {
+  auto needed_instruction =
+      FindInstruction(message_.needed_instruction(), ir_context);
+
+  uint32_t needed_index =
+      GetMemoryScopeInOperandIndex(needed_instruction->opcode());
+
+  needed_instruction->SetInOperand(needed_index,
+                                   {message_.memory_scope_new_value_id()});
+}
+
+uint32_t TransformationChangingMemoryScope::GetMemoryScopeInOperandIndex(
+    SpvOp opcode) {
+  switch (opcode) {
+    case SpvOpAtomicLoad:
+    case SpvOpAtomicStore:
+    case SpvOpAtomicExchange:
+    case SpvOpAtomicIIncrement:
+    case SpvOpAtomicIDecrement:
+    case SpvOpAtomicIAdd:
+    case SpvOpAtomicISub:
+    case SpvOpAtomicSMin:
+    case SpvOpAtomicUMin:
+    case SpvOpAtomicSMax:
+    case SpvOpAtomicUMax:
+    case SpvOpAtomicAnd:
+    case SpvOpAtomicOr:
+    case SpvOpAtomicXor:
+    case SpvOpAtomicFlagTestAndSet:
+    case SpvOpAtomicFlagClear:
+    case SpvOpAtomicFAddEXT:
+    case SpvOpAtomicCompareExchange:
+    case SpvOpAtomicCompareExchangeWeak:
+      return 1;
+
+    default:
+      assert(false);
+      return -1;
+  }
+}
+
+bool TransformationChangingMemoryScope::IsAtomicInstruction(SpvOp opcode) {
+  switch (opcode) {
+    // Atomic Instructions.
+    case SpvOpAtomicLoad:
+    case SpvOpAtomicStore:
+    case SpvOpAtomicExchange:
+    case SpvOpAtomicCompareExchange:
+    case SpvOpAtomicCompareExchangeWeak:
+    case SpvOpAtomicIIncrement:
+    case SpvOpAtomicIDecrement:
+    case SpvOpAtomicIAdd:
+    case SpvOpAtomicISub:
+    case SpvOpAtomicSMin:
+    case SpvOpAtomicUMin:
+    case SpvOpAtomicSMax:
+    case SpvOpAtomicUMax:
+    case SpvOpAtomicAnd:
+    case SpvOpAtomicOr:
+    case SpvOpAtomicXor:
+    case SpvOpAtomicFlagTestAndSet:
+    case SpvOpAtomicFlagClear:
+    case SpvOpAtomicFAddEXT:
+      return true;
+
+    default:
+      return false;
+  }
+}
+
+bool TransformationChangingMemoryScope::IsValidScope(
+    SpvScope new_memory_scope_value, SpvScope old_memory_scope_value) {
+  switch (new_memory_scope_value) {
```

Comment:
If old_memory_scope_value is not one of the values below, we should return false.

Reply:
Done.